### PR TITLE
Add arm32v7 variant of Stretch images

### PR DESCRIPTION
**Pull request by richlander**
- We want to add ARM32v7 images to fix https://github.com/dotnet/dotnet-docker/issues/223
- Based on guidance from Docker folks, we will integrate the new images in our existing dotnet and dotnet-nightly repos. See: https://github.com/moby/moby/issues/33759.
- We will make significant use of Docker multi-arch (MA) for dotnet images.
- At a later point, we can likely make use of multi-arch for Debian dependencies (when MA shows up for Debian base images).

This PR is not done -- left to do:
- [x] Validate correctness of the Dockerfiles
- [x] Update README to include arm32 appropriately.
- [x] Generate appropriate MA tags/manifests.

### CODE REVIEW

**Comment by MichaelSimons** [comment]
Would it make sense to put the existing Dockerfiles in an architecture directory so that the directory structure is consistent across platforms?

/2.0
    /runtime
        /stretch
            /amd64
            /arm32v7

**Comment by richlander** [comment]
Your formatting got messed up, but I'm pretty sure I know what you meant. 

I spent some time looking at our README to see how the GitHub source links would show up per image. Upon further consideration, I like your suggestion better. My suggestion falls a little too much in the "clever" category by compressing the folder structure.

**Comment by richlander** [comment]
We could add an amd64 directory to the jessie directory for symmetry or leave as-is to keep things simple. Thoughts?

**Comment by richlander** [comment]
Just tested ASP.NET Core preview 2 within these Docker images on raspbian. The app was `dotnet new mvc`.

docker

**Comment by richlander** [comment]
What is left @MichaelSimons? I think we are getting close to landing this PR.

**Comment by MichaelSimons** [comment]
@richlander, my preference would be to add the amd64 dir for Jessie as well in order to have symmetry.

**Comment by MichaelSimons** [comment]
@richlander - Regarding your question about the remaining work.  
- [ ] Machines to build arm images in official builds.  I have a request out to DDIT on this.
- [ ] The image-builder tool used in official builds needs a [minor tweak to support ARM builds](https://github.com/dotnet/docker-tools/issues/19).  I have the changes locally.
- [ ] Once image-builder supports ARM, the new version needs to picked up and then the [manifest.json](https://github.com/dotnet/dotnet-docker/blob/master/manifest.json) needs to be updated so that official builds will build the MA tags for the ARM images.
- [ ] Unit Tests should be updated to cover the new runtime images.

**Comment by richlander** [comment]
@MichaelSimons I am getting back to this. We should re-up on where we are at. I was looking at the manifest.json. It would be good to get some guidance on how like you'd like to pivot architecture within the format. I can see a couple options.

I just did a `git pull --rebase` to make this branch current.

Any progress on the work items you mentioned above?

**Comment by MichaelSimons** [comment]
@richlander, I just made https://github.com/dotnet/docker-tools/pull/28 to modify the image-builder tool to support ARM.  In the process of testing my change, I updated the manifest locally.  Once the PR is merged I will check-in my manifest changes.   The blocking issue right now is to get machines to do official builds on.  I will be pushing on this over the next few days to get us unblocked.

**Comment by richlander** [comment]
Cool. I was wondering how you would model another arch. I haven't thought about it deeply, but that looks good. I could see that the model wasn't quite going to survive as-is. Thanks!

Would be great to get unblocked! Thanks.

**Comment by MichaelSimons** [comment]
@richlander, @dagood I think this PR is ready for review.  I plan on writing a CI test for the ARM images but see that as a separate PR.

**Comment by MichaelSimons** [comment]
If you want to try out the multi-arch arm32 tags, I used the build infrastructure changes within this PR to build and push the images to msimons/dotnet-nightly

File: 2.0/runtime-deps/stretch/arm32v7/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        libcurl3 \
+        libgcc1 \
+        libgssapi-krb5-2 \
+        libicu57 \
+        liblttng-ust0 \
+        libssl1.0.2 \
+        libstdc++6 \
+        libunwind8 \
+        libuuid1 \
+        zlib1g \
```
Comment (dagood) [inline 2.0/runtime-deps/stretch/arm32v7/Dockerfile]: Weird way to end a file, is this missing the cleanup?

Comment (MichaelSimons) [inline 2.0/runtime-deps/stretch/arm32v7/Dockerfile]: Yes it is - looks like the `&& rm -rf /var/lib/apt/lists/*` got chopped off.

Comment (MichaelSimons) [inline 2.0/runtime-deps/stretch/arm32v7/Dockerfile]: Fixed

**Comment by richlander** [comment]
LGTM

**Comment by alexellis** [comment]
How close or possible is an SDK image for ARM 32-bit? I saw there is a Runtime - but the source needs to be built too.

**Comment by MichaelSimons** [comment]
@alexellis - The .NET Core SDK does not run on the Linux + ARM32 configuration.  You will need to build your application on X64, copy the binaries to an ARM32 device and then build your Docker image.  Alternatively you can make use of a multi-stage Dockerfile on Window 10 or macOS.  The dotnet-docker-samples include a [sample](https://github.com/dotnet/dotnet-docker-samples/tree/master/dotnetapp-prod#build-on-windows-or-macos-and-run-the-sample-with-docker-on-linux--arm32-raspberry-pi) of this flow.  If you have feedback on the sample please feel open an issue in the [dotnet-docker-samples repo](https://github.com/dotnet/dotnet-docker-samples/issues).

**Comment by alexellis** [comment]
From a workflow/CI perspective this is disappointing. I work with the RPi platform + Docker a lot as a Docker Captain and .NET developer at work.

Is this a temporary stumbling block or a permanent design decision? What about 64-bit ARM? Packet.net for instance have enterprise-grade Cavium servers which are very generously spec'd.

**Comment by richlander** [comment]
I wrote up the situation here: https://github.com/dotnet/announcements/issues/29.

So, no, not a permanent decision, but also not our highest priority to solve at the moment. The rational given (performance) wouldn't apply to ARM64, so I would expect a different choice there.

Our first focus is reducing the size of the SDK generally. Once we have that in-hand, then making a good experience on ARM32 should be easier.

**Pull request merged**